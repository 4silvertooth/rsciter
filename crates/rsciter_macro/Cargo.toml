[package]
name = "rsciter_macro"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[dependencies]
syn = { version = "1.0", features = ["full"] }
proc-macro2 = "1.0"
quote = "1.0"
proc-macro-error = "1.0"

[dev-dependencies]
trybuild = "1"
rsciter = { path = "../rsciter" }