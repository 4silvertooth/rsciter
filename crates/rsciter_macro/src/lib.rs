use proc_macro::TokenStream;
use proc_macro2::{Ident, TokenStream as TokenStream2};
use proc_macro_error::{proc_macro_error, ResultExt};
use quote::quote;
use syn::spanned::Spanned;

pub(crate) mod items;
pub(crate) mod sciter_mod;

#[proc_macro_attribute]
pub fn transparent(_attr: TokenStream, input: TokenStream) -> TokenStream {
    input
}

#[proc_macro_error]
#[proc_macro_attribute]
pub fn xmod(attr: TokenStream, input: TokenStream) -> TokenStream {
    let data = xmod_impl(attr.into(), input.into()).unwrap_or_abort();
    data.into()
}

fn xmod_impl(attr: TokenStream2, input: TokenStream2) -> syn::Result<TokenStream2> {
    match syn::parse2::<syn::ItemMod>(input) {
        Ok(m) if m.content.is_none() => Err(syn::Error::new(
            m.span(),
            "the #[rsciter::xmod] attribute can only be applied to an inline module!",
        )),
        Err(e) => Err(syn::Error::new(
            e.span(),
            format!("{e}: the #[rsciter::xmod] attribute can only be applied to an inline module!"),
        )),
        Ok(mut module) => {
            let mut struct_name = attr.to_string();
            if struct_name.is_empty() {
                struct_name = module.ident.to_string();
                // struct and module names are in the same namespace,
                // have to rename the module to use its name
                module.ident = Ident::new(&format!("{}_mod", &module.ident), module.ident.span());
            }

            let info = sciter_mod::SciterMod::new(&module, struct_name)?;
            let provider_struct_name = info.ident();
            let vis = info.visibility();
            let (names, calls, implementations) = info.methods();

            Ok(quote!(
                #[allow(non_snake_case)]
                #[allow(dead_code)]
                #module // TODO: remove attrs like #[tranparent]

                #vis struct #provider_struct_name;

                impl ::rsciter::XFunctionProvider for #provider_struct_name {
                    fn call(&mut self, name: &str, args: &[::rsciter::Value]) -> ::rsciter::Result<Option<::rsciter::Value>> {
                        match name {
                            #( #names => #calls, )*
                            _ => Err(::rsciter::Error::ScriptingNoMethod(name.to_string())),
                        }
                    }
                }

                impl #provider_struct_name {
                    #( #implementations )*
                }
            ))
        }
    }
}
