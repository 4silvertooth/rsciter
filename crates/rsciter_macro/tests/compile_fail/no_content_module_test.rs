use rsciter_macro::xmod;

#[xmod]
mod no_content;

fn main() {}