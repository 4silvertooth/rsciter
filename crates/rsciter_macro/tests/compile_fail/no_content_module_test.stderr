error[E0658]: non-inline modules in proc macro input are unstable
 --> tests/compile_fail/no_content_module_test.rs:4:1
  |
4 | mod no_content;
  | ^^^^^^^^^^^^^^^
  |
  = note: see issue #54727 <https://github.com/rust-lang/rust/issues/54727> for more information

error: the #[rsciter::xmod] attribute can only be applied to an inline module!
 --> tests/compile_fail/no_content_module_test.rs:4:1
  |
4 | mod no_content;
  | ^^^
